pkg/helmdeploy/write.go

```diff
@@ -0,0 +1,212 @@
+			}
+		}
+	}
+
+	return nil
+}
+
+func renderHelmChart(opts WriteOptions, downstream string, helmChart *kotsv1beta2.HelmChart) (string, error) {
+	cfg := &action.Configuration{
+		Log: opts.Log.Debug,
+	}
+	client := action.NewInstall(cfg)
+	client.DryRun = true
+	client.ReleaseName = helmChart.GetReleaseName()
+	client.Replace = true
+	client.ClientOnly = true
+	client.IncludeCRDs = true
+
+	chartDir := path.Join(opts.HelmDir, helmChart.GetDirName())
+
+	client.Namespace = helmChart.Spec.Namespace
+	if client.Namespace == "" {
+		client.Namespace = util.PodNamespace
+	}
+
+	chartPath := path.Join(chartDir, fmt.Sprintf("%s-%s.tgz", helmChart.Spec.Chart.Name, helmChart.Spec.Chart.ChartVersion))
+	valuesPath := path.Join(chartDir, "values.yaml")
+
+	chartRequested, err := loader.Load(chartPath)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to load chart")
+	}
+
+	if req := chartRequested.Metadata.Dependencies; req != nil {
+		if err := action.CheckDependencies(chartRequested, req); err != nil {
+			return "", errors.Wrap(err, "failed dependency check")
+		}
+	}
+
+	values, err := chartutil.ReadValuesFile(valuesPath)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to read values file")
+	}
+
+	rel, err := client.Run(chartRequested, values)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to run helm install")
+	}
+
+	var manifests bytes.Buffer
+	fmt.Fprintln(&manifests, strings.TrimSpace(rel.Manifest))
+	for _, m := range rel.Hooks {
+		fmt.Fprintf(&manifests, "---\n# Source: %s\n%s\n", m.Path, m.Manifest)
+	}
+	// add hooks
+	for _, m := range rel.Hooks {
+		fmt.Fprintf(&manifests, "---\n# Source: %s\n%s\n", m.Path, m.Manifest)
+	}
+
+	renderedPath := path.Join(opts.RenderedDir, downstream, "helm", helmChart.GetDirName())
+	if err := os.MkdirAll(renderedPath, 0744); err != nil {
+		return "", errors.Wrap(err, "failed to create rendered path")
+	}
+
+	err = os.WriteFile(filepath.Join(renderedPath, "all.yaml"), manifests.Bytes(), 0644)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to write all.yaml")
+	}
+
+	return renderedPath, nil
+}
+
+// processImages will pull any public images from online and copy them to the configured private registry
+func processImages(opts WriteOptions, renderedPath string) error {
+	if !opts.ProcessImageOptions.RewriteImages || opts.ProcessImageOptions.AirgapRoot != "" {
+		return nil
+	}
+
+	var dockerHubRegistryCreds registry.Credentials
+	dockerhubSecret, _ := registry.GetDockerHubPullSecret(opts.Clientset, util.PodNamespace, opts.ProcessImageOptions.Namespace, opts.ProcessImageOptions.AppSlug)
+	if dockerhubSecret != nil {
+		dockerHubRegistryCreds, _ = registry.GetCredentialsForRegistryFromConfigJSON(dockerhubSecret.Data[".dockerconfigjson"], registry.DockerHubRegistryName)
+	}
+
+	_, err := midstream.RewriteBaseImages(opts.ProcessImageOptions, renderedPath, opts.KotsKinds, opts.KotsKinds.License, dockerHubRegistryCreds, opts.Log)
```

Comment:
I don't love the name of this function... I wonder if we can refactor some of that code at some point, since neither `midstream` nor `RewriteBaseImages` make sense in this context.

Reply:
seems we might have to do this refactor to get all of those other functions into a common `apparchive` pkg.  I'll have a go at this...